### PR TITLE
iox-#2224 Fix const value assignment in optional

### PR DESCRIPTION
**Pull request by tdouglas-latai**
Fixes the forwarding reference type in `iox::optional::operator=` when performing a direct assignment of the underlying value such that assigning `const` values is supported.

## Pre-Review Checklist for the PR Author

1. [ ] ~Add a second reviewer for complex new features or larger refactorings~
1. [x] Code follows the coding style of [CONTRIBUTING.md][contributing]
1. [x] Tests follow the [best practice for testing][testing]
1. [x] Changelog updated [in the unreleased section][changelog] including API breaking changes
1. [x] Branch follows the naming format (`iox-123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed (except `task-list-completed`)
1. [x] All touched (C/C++) source code files from `iceoryx_hoofs` are added to `./clang-tidy-diff-scans.txt`
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php
[contributing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/CONTRIBUTING.md#coding-style
[testing]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/concepts/best-practice-for-testing.md
[changelog]: https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/release-notes/iceoryx-unreleased.md

## Checklist for the PR Reviewer

- [x] Commits are properly organized and messages are according to the guideline
- [x] Code according to our coding style and naming conventions
- [x] Unit tests have been written for new behavior
- [x] Public API changes are documented via doxygen
- [x] Copyright owner are updated in the changed files
- [x] All touched (C/C++) source code files from `iceoryx_hoofs` have been added to `./clang-tidy-diff-scans.txt`
- [x] PR title describes the changes

## Post-review Checklist for the PR Author

1. [x] All open points are addressed and tracked via issues

## References

- Closes [#2224](https://github.com/eclipse-iceoryx/iceoryx/issues/2224)

### CODE REVIEW

File: iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl [hunk 1]
```diff
@@ -169,11 +169,11 @@ optional<T>::operator=(U&& newValue) noexcept
 {
     if (m_hasValue)
     {
-        value() = std::forward<T>(newValue);
```
Comment (elfenpiff) [inline iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:172]: I think we have to revert this change. Maybe the solution is to add explicit "copy/move" assignment operators for the underlying value like:

```cxx
// in inl file
template <typename T>
inline optional<T>& optional<T>::operator=(const T& newValue) noexcept
{
    if (m_hasValue)
    {
        value() = newValue;
    }
    else
    {
        construct_value(newValue);
    }
    return *this;
}

template <typename T>
inline optional<T>& optional<T>::operator=(T&& newValue) noexcept
{
    if (m_hasValue)
    {
        value() = std::move(newValue);
    }
    else
    {
        construct_value(std::move(newValue));
    }
    return *this;
}
```

and add the definition to the hpp file:
```cxx
optional& operator=(const T& newValue) noexcept;
optional& operator=(T&& newValue) noexcept;
```

Comment (tdouglas-latai) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:172]: Taking a look at https://github.com/eclipse-iceoryx/iceoryx/blob/1e17b23d50e7ac2243f43969d90e7f5e8b60afca/tools/introspection/source/introspection_app.cpp#L796 where one of the errors comes from, it seems like a side effect of the current behavior is to implicitly move when assigning a non-const reference.

The `std::forward<T>` turns into `T&&`, while `std::forward<U>` with `U = T&` is equivalent to `T&`.  This matches the copy constructor of `Sample` (which is deleted) rather than the move constructor.  It seems like this is not desired behavior (implicitly converting an l-value to r-value reference).  I can fix it here by explicitly doing `memPoolSample = std::move(sample)`, though I'm not sure what implications this might have elsewhere in Iceoryx or for other users.

Comment (tdouglas-latai) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:172]: This was the only file I found with issues when building locally, so I pushed an update to try again.  Let me know how you'd like to proceed if CI passes.

Comment (elfenpiff) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:172]: We should realize this quickly when this was the wrong approach, so I would go for it and test the introspection app afterwards. If it still shows all the mempools we should be fine.

In my opinion, we are with iceoryx at a state with testing and the CI that we can perform fearless refactoring and rely on the CI.

Comment (tdouglas-latai) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:172]: I ran the introspection client with the changes, along with just `iox-roudi` and I see this output: 
image

Comment (elfenpiff) [reply iceoryx_hoofs/vocabulary/include/iox/detail/optional.inl:172]: Mempool looks exactly like it should. Nice.

**Comment by tdouglas-latai** [comment]
Hmm, it seems like the ASAN build may have crashed or something?
```
2024-03-13T19:15:11.4614668Z [----------] Global test environment tear-down
2024-03-13T19:15:11.4942698Z [==========] 21 tests from 2 test suites ran. (41 ms total)
2024-03-13T19:15:11.4943362Z [  PASSED  ] 21 tests.
2024-03-13T19:15:11.5740157Z AddressSanitizer:DEADLYSIGNAL
2024-03-13T19:15:11.5740803Z AddressSanitizer:DEADLYSIGNAL
2024-03-13T19:15:11.5741310Z AddressSanitizer:DEADLYSIGNAL
2024-03-13T19:15:11.5741823Z AddressSanitizer:DEADLYSIGNAL
2024-03-13T19:15:11.5742317Z AddressSanitizer:DEADLYSIGNAL
2024-03-13T19:15:11.5742807Z AddressSanitizer:DEADLYSIGNAL
2024-03-13T19:15:11.5743292Z AddressSanitizer:DEADLYSIGNAL
2024-03-13T19:15:11.5743782Z AddressSanitizer:DEADLYSIGNAL
```
Is there something I can do about that? The `DEADLYSIGNAL` spew is so much I have to stop the raw log page from loading before it's responsive :sweat_smile:

**Comment by elfenpiff** [comment]
@tdouglas-latai maybe it was a CI hiccup, I will rerun it again. As far as I can see, the STDLIB_tests did run successfully and then the address sanitizer failed. Without any reason ...

**Comment by tdouglas-latai** [comment]
Hmm, seems to have failed in a similar way the second time.  I ran `./tools/ci/build-test-ubuntu-with-sanitizers.sh gcc asan` on my local machine and it seemed to pass without issue.  Maybe rerun a third time?

**Comment by tdouglas-latai** [comment]
@elfenpiff, I tried running those commands and did get some failures during build with the compiler being killed, even got those errors on `master`, but I didn't get the same error as you or CI.  I changed the build jobs to 4 and it passed on my local machine using this branch and `master`.  Do you have any idea from your output if this could be caused by the proposed changes here?

**Comment by elfenpiff** [comment]
@tdouglas-latai I can reproduce it with your branch but not on master. So it is definitely caused by your changes, but I have no idea what the issue here is. Especially, since it is crashing at the end of a successful test run and the test run varies as well, sometimes after the STDLIB tests, sometimes after the hoofs_moduletests.

I tend to say this is some bug within the address sanitizer, but we cannot ignore this, otherwise we have a failing CI on master. We either need to confirm that it is in fact an address sanitizer bug and disable the CI target, or we have to find the issue on our side

**Comment by tdouglas-latai** [comment]
Yes, definitely can't ignore the CI failure, I'll play with it more to try and replicate it locally.

**Review by elfenpiff — APPROVED** [review]
The CI issue seems to be gone - most likely some ubuntu problem.

**Comment by tdouglas-latai** [comment]
Oh, sweet, thanks!
